android/src/main/java/com/oktareactnative/HttpClientImpl.java

```diff
@@ -12,166 +12,161 @@
 import com.okta.oidc.net.ConnectionParameters;
 import com.okta.oidc.net.OktaHttpClient;
-import com.okta.oidc.net.request.TLSSocketFactory;
 
-import java.io.DataOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
-import java.net.HttpURLConnection;
-import java.net.URL;
-import java.security.KeyManagementException;
-import java.security.NoSuchAlgorithmException;
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
 
-import javax.net.ssl.HttpsURLConnection;
-
-import static com.okta.oidc.net.ConnectionParameters.USER_AGENT;
+import okhttp3.Call;
+import okhttp3.Callback;
+import okhttp3.FormBody;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
 
 public class HttpClientImpl implements OktaHttpClient {
     private final String userAgentTemplate;
     private final int connectTimeoutMs;
     private final int readTimeoutMs;
-
-    private HttpURLConnection mUrlConnection;
+    protected static OkHttpClient sOkHttpClient;
+    protected volatile Call mCall;
+    protected Response mResponse;
+    protected Exception mException;
 
     HttpClientImpl(String userAgentTemplate, int connectTimeoutMs, int readTimeoutMs) {
         this.userAgentTemplate = userAgentTemplate;
         this.connectTimeoutMs = connectTimeoutMs;
         this.readTimeoutMs = readTimeoutMs;
     }
 
-    /*
-     * TLS v1.1, v1.2 in Android supports starting from API 16.
-     * But it enabled by default starting from API 20.
-     * This method enable these TLS versions on API < 20.
-     * */
-    @SuppressLint("RestrictedApi")
-    private void enableTlsV1_2(HttpURLConnection urlConnection) {
-        try {
-            ((HttpsURLConnection) urlConnection)
-                    .setSSLSocketFactory(new TLSSocketFactory());
-        } catch (NoSuchAlgorithmException | KeyManagementException e) {
-            throw new RuntimeException("Cannot create SSLContext.", e);
-        }
-    }
-
     private String getUserAgent() {
         String sdkVersion = "okta-oidc-android/" + BuildConfig.VERSION_NAME;
         return userAgentTemplate.replace("$UPSTREAM_SDK", sdkVersion);
     }
 
-    protected HttpURLConnection openConnection(URL url, ConnectionParameters params)
-            throws IOException {
-        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
-        if (mUrlConnection instanceof HttpsURLConnection &&
-                Build.VERSION.SDK_INT <= Build.VERSION_CODES.LOLLIPOP) {
-            enableTlsV1_2(mUrlConnection);
+    protected Request buildRequest(Uri uri, ConnectionParameters param) {
+        if (sOkHttpClient == null) {
+            sOkHttpClient = new OkHttpClient.Builder()
+                    .connectTimeout(connectTimeoutMs, TimeUnit.MILLISECONDS)
+                    .readTimeout(readTimeoutMs, TimeUnit.MILLISECONDS)
+                    .build();
         }
```

Comment:
It's perfectly okay if this code gets called by multiple callers at the same time. okHttpClient doesn't need to be singleton for correctness, but only for performance reasons.